CMake: Add CMake to mbed-psa greentea tests

hazzlim opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
- Changed ```mbed_greentea_add_test``` macro to require TEST_SOURCES for all files, removing assumption of main.cpp
- Refactored all pre-existing greentea tests to add main.cpp using TEST_SOURCES
- Added a CMakeLists.txt for each test suite in Mbed OS PSA. 
- Moved /pal and /val directories into a separate /test_abstraction_layer directory and combined them into one CMake target, which is linked to by the test suites that require them.

Fixes #14680

#### Impact of changes <!-- Optional -->

<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

None.

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
None.

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
None. 

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR

Manual testing for pre-existing greentea tests:

All pre-existing tests built successfully using below commands, or fail as expected due to a [NOT_SUPPORTED] error.
    
Manual testing for PSA tests:
Successfully able to build the test suite with below manual steps:
- ```mbedtools configure -t <toolchain> -m <target> --mbed-os-path <path-to-mbed-os> --app-config TESTS/configs/experimental.json``` (requires [this PR](https://github.com/ARMmbed/mbed-tools/pull/292))
- ```cd cmake_build/<target>/develop/<toolchain>/ && cmake ../../../.. -G Ninja -DMBED_TEST_LINK_LIBRARIES=mbed-os && cmake --build .```
- ```mbedhtrun -f <path/to/image> -p <port> -d <mount>```

All built successfully with both GCC_ARM and ARMC6, and run successfully on target K66F, except:
- entropy_inject: can only run with MBEDTLS_ENTROPY_NV_SEED defined; build fails with ```#error [NOT_SUPPORTED] PSA entropy injection tests can run only with MBEDTLS_ENTROPY_NV_SEED enabled.``` (Build successful if the macro is defined, but test crashes when run on K66F)
- compliance_its/test_s003: can only run with PS_ALLOW_ENTIRE_STORAGE_FILL defined; build fails with ```#error [NOT_SUPPORTED] Test is too long for CI, thus always fails on timeout.``` (Build successful if the macro is defined, and test runs to {{result;success}})
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

@Patater @LDong-Arm

----------------------------------------------------------------------------------------------------------------

hazzlim committed:
CMake: Add support for PSA attestation test

The greentea test for mbed-psa attestation can now be built with CMake.

hazzlim committed:
CMake: Add support for PSA crypto_init test

The greentea test for mbed-psa crypto_init can now be built with CMake.

hazzlim committed:
CMake: Add support for PSA its_ps test

The greentea test for mbed-psa its_ps can now be built with CMake.

hazzlim committed:
CMake: Add mbed-psa-tal (Test Abstraction Layers)

Move /val and /pal directories into /test_abstraction_layers directory
and combine into one CMake target, mbed-psa-tal. Moved into seperate
directory in order to have own CMakeLists.txt, rather than adding to
/TARGET_MBED_PSA_SRV CMake file.

hazzlim committed:
CMake: Add support for compliance_attestation test

Greentea test for PSA compliance_attestation can now build with CMake.

hazzlim committed:
CMake: Add support for PSA compliance_its tests

PSA compliance_its tests can now be built with CMake.

hazzlim committed:
CMake: Add support for PSA entropy_inject test

The greentea test for PSA entropy_inject can now be built with CMake.
Note: requires MBEDTLS_ENTROPY_NV_SEED enabled, so not tested on target.

LDong-Arm commented on platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_MBED_PSA_SRV/TESTS/attestation/test/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+cmake_minimum_required(VERSION 3.19.0 FATAL_ERROR)
+
+set(MBED_PATH ${CMAKE_CURRENT_SOURCE_DIR}/../../../../../../../ CACHE INTERNAL "")
+set(TEST_TARGET mbed-platform-feature-psa-attestation)
```
No need to have "feature" in the CMake target name. The path names FEATURE_xxx are purely an Mbed CLI 1 "magic", to compile or skip a directory based on configuration.

hazzlim replied:
Have removed 'feature' from all test target names 👍

LDong-Arm commented on platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_MBED_PSA_SRV/TESTS/compliance_attestation/test_a001/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+project(${TEST_TARGET})
+
+mbed_greentea_add_test(
+    TEST_NAME ${TEST_TARGET}
```
Nitpicking: 
```suggestion
    TEST_NAME
        ${TEST_TARGET}
```
to be consistent with the next few lines

hazzlim replied:
Fixed for all test targets with multiple args for mbed_greentea_add_test()

LDong-Arm commented on platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_MBED_PSA_SRV/TESTS/crypto_init/test/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+project(${TEST_TARGET})
+
+mbed_greentea_add_test(
+    TEST_NAME ${TEST_TARGET})
```
```suggestion
mbed_greentea_add_test(TEST_NAME ${TEST_TARGET})
```
This is not very important, but the style in Mbed OS is either have the whole statement in a single line, or have the closing `)` in its own line if multiple lines are used.

hazzlim replied:
Fixed for this and similar targets 👍

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -22,39 +22,39 @@ include(${CMAKE_CURRENT_LIST_DIR}/app.cmake)
-            main.cpp
+            ${MBED_GREENTEA_TEST_MAIN} 
             ${MBED_GREENTEA_TEST_SOURCES}
     )
```
For other reviewers' reference only:

The old assumption of `main.cpp` is not always true - we shouldn't hardcode any file names here or prevent a test from using C or having a different name. In my opinion, two better designs could be:
* Option 1: Default to `main.cpp` only when the optional `TEST_SOURCES` is unset. If `TEST_SOURCES` is set by a test, the test should include the main program in this parameter.
* Option 2: Don't assume `main.cpp`, and make `TEST_SOURCES` compulsory.

Either option would require changing CMake definitions of existing tests which are many. So to avoid doing that now, a new parameter `TEST_MAIN` is added in this PR, though obviously having both `TEST_MAIN` and `TEST_SOURCES` is largely redundant. Maybe this can be refactored in the future.

Patater replied:
Making TEST_SOURCES compulsory sounds like a good approach to me. If we want to do this, we should refactor now as part of this work. I don't think it's an important enough of a change we'd get to it later.

hazzlim replied:
> Making TEST_SOURCES compulsory sounds like a good approach to me. If we want to do this, we should refactor now as part of this work. I don't think it's an important enough of a change we'd get to it later.

I can update my PR with this refactoring, if that's what makes sense to everyone else

LDong-Arm replied:
Given that there're a very large number of tests, maybe we can create a separate PR to do that, and based this PR on that?

The change itself should be straightforward, i.e. use
```
grep -rnI mbed_greentea_add_test .
```
to find all occurrences.

hazzlim replied:
@LDong-Arm @Patater I have made TEST_SOURCES compulsory and updated all greentea tests accordingly. The changes are rebased into this PR.

LDong-Arm reviewed (commented):
@hazzlim Thanks for the PR, it looks pretty good to me! I will try a few tests locally.

LDong-Arm commented on hal/tests/TESTS/mbed_hal/flash/functional_tests/CMakeLists.txt:
```diff
@@ -3,11 +3,16 @@
 cmake_minimum_required(VERSION 3.19.0 FATAL_ERROR)
 
-set(MBED_PATH ${CMAKE_CURRENT_SOURCE_DIR}/../../../../.. CACHE INTERNAL "")
+set(MBED_PATH ${CMAKE_CURRENT_SOURCE_DIR}/../../../../../.. CACHE INTERNAL "")
```
Good catch, so you actually validated them! 👍

Patater replied:
Could you move this into its own commit please? It's not really related to making `TEST_SOURCES` mandatory.

LDong-Arm commented on hal/tests/TESTS/mbed_hal_fpga_ci_test_shield/CMakeLists.txt:
```diff
@@ -1,13 +0,0 @@
-
-project(${TEST_TARGET})
-
-mbed_greentea_add_test(TEST_NAME ${TEST_TARGET})
```
Good catch

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -26,35 +26,29 @@ macro(mbed_greentea_add_test)
-    add_executable(${TEST_NAME})
-
-    # Explicitly enable BUILD_TESTING until CTest is added to the Greentea client
+    add_executable(${MBED_GREENTEA_TEST_NAME}) # Explicitly enable BUILD_TESTING until CTest is added to the Greentea client
```
The comment about BUILD_TESTING refers to the next line and not this one.

hazzlim replied:
Thank you I didn't see this! I have updated with this change, & also fixed a bracket in mbed_greentea.cmake too

LDong-Arm reviewed (commented):
Thanks @hazzlim, it looks almost perfect to me - nice improvement to `mbed_greentea_add_test`.
Just one remark on a line of comment.

LDong-Arm reviewed (dismissed):
LGTM

Patater commented on hal/tests/TESTS/mbed_hal_fpga_ci_test_shield/CMakeLists.txt:
```diff
@@ -1,13 +0,0 @@
-# Copyright (c) 2020 ARM Limited. All rights reserved.
-# SPDX-License-Identifier: Apache-2.0
```
Could we remove this file in its own separate commit please, ensuring the commit message explains the reasoning.

hazzlim replied:
Done 👍

Patater commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -35,26 +35,23 @@ macro(mbed_greentea_add_test)
     add_subdirectory(${MBED_PATH} build)
 
-    add_executable(${TEST_NAME})
+    add_executable(${MBED_GREENTEA_TEST_NAME})
```
The "MBED_GREENTEA_TEST_NAME" change should also go in its own commit.

hazzlim replied:
Done 👍

Patater commented on platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_MBED_PSA_SRV/TESTS/compliance_attestation/test_a001/main.c:
```diff
@@ -20,6 +20,7 @@
 #if !defined(MBED_CONF_RTOS_PRESENT)
 #error [NOT_SUPPORTED] PSA compliance attestation test cases require RTOS to run.
 #else
+
```
We should leave this as-is, to avoid diff noise. This isn't necessary to "Add support for compliance_attestation test", so it shouldn't be part of that commit if we even want to make the change at all (I don't think we care to).

hazzlim replied:
Removed !

hazzlim committed:
CMake: fix incorrect relative path for MBED_PATH

MBED_PATH set to /platform directory incorrectly. Updated to set to
/mbed-os directory.

hazzlim committed:
CMake: Remove unnecessary CMakeLists.txt

CMakeLists.txt file in /hal/tests/TESTS/mbed_hal_fpga_ci_test_shield
directory was non-functional as it used the greentea_add_test macro,
which expects a main.cpp in current directory, but no main.cpp exists
there. I checked with @rajkan01 who confirmed that this CMake file is
serving no purpose and is there erroneously. All tests in subdirectories
of this directory have their own CMakeLists.txt that successfully build
them.

0xc0170 commented on connectivity/FEATURE_BLE/source/cordio/TESTS/cordio_hci/driver/CMakeLists.txt:
```diff
@@ -11,6 +11,11 @@ include(${MBED_PATH}/tools/cmake/mbed_greentea.cmake)
 mbed_greentea_add_test(
-    TEST_NAME ${TEST_TARGET}
-    TEST_REQUIRED_LIBS mbed-ble mbed-events
+    TEST_NAME
```
The style change - was it done by a tool or just follow CMake docs style?

hazzlim replied:
Was done to follow style pointers from Lingkai for previous commits

hazzlim replied:
Possibly a bit of an over-interpretation - is it best as it was?

LDong-Arm replied:
This is the kind of indentation we have when calling a macro function with parameters, so it looks good to me. I guess what @0xc0170 means is whether it was done manually or by some tool, and the former is the case.

rwalton-arm replied:
We should set-up `cmake-format` as a pre-commit check and specify the style we want to use. It would be great to run the `cmake-format` check as part of the CI. That way we will auto-enforce the style and it will always be consistent. We can add a new issue to add `cmake-format` to the CI in a separate PR.

hazzlim committed:
CMake: Remove alias for TEST_NAME argument

The mbed_greentea_add_test macro previously set a variable in order to
use the un-prefixed TEST_NAME to refer to the argument in the macro
body. Whilst pair-programming with LDong, this was identified and
determined to be unecessary (maybe it was a failed attempt to fix
something, that was never reversed?) and so it has been removed.

hazzlim committed:
CMake: Require TEST_SOURCES in greentea CMake file

Assumption that greentea test file is always named main.cpp is
incorrect. Updated mbed_greentea_add_test() macro to make TEST_SOURCES
parameter compulsory, which is used to specify greentea test
file(s). This allows tests to use C, or have a different name.
Therefore also updated all pre-existing greentea test CMake files to
explicity add main.cpp to TEST_SOURCES.

Patater commented on hal/tests/TESTS/mbed_hal/flash/functional_tests/CMakeLists.txt:
```diff
@@ -3,11 +3,16 @@
 project(${TEST_TARGET})
 
-mbed_greentea_add_test(TEST_NAME ${TEST_TARGET})
+mbed_greentea_add_test(
```
This change here shouldn't be part of "CMake: fix incorrect relative path for MBED_PATH". It should be part of "CMake: Require TEST_SOURCES in greentea CMake file" instead.

hazzlim replied:
Ah snap I did not catch that! Done now 👍

Patater reviewed (approved):
LGTM
